### ISSUE TITLE
Rule: async callback should be paired with return

### ISSUE THREAD

**Issue by puzrin** [eslint/eslint#994]
In `node.js` typical error is writing something like this:

```javascript
if (err) {
  callback(err);
}
```

Right cases are:

1.

```javascript
if (err) {
  callback(err);
  return;
}
```

2.

```javascript
if (err) return callback (err);
```

Imho, this can be automated. Preliminary conditions to detect callback

- keywords `callback`, `cb`, `next` in function name
- 0-2 params

**Comment by fakewaffle** [comment]
Any update on this?

**Comment by ilyavolodin** [comment]
Issue is in accepted state. So anyone can submit a pull request for it. New rules are low priority for core team right now, so if you would like to see this rule, you might consider submitting a pull request.

**Comment by xjamundx** [comment]
It sounds like @puzrin proposed it, @jrajav had some concerns and nobody has yet taken it up yet.

I built something like this for my team, at least some variation of it. It's amazingly hard to get right and does not really work perfectly, but I'd be happy to share what I came up with as a strawman in the coming days...

**Comment by xjamundx** [comment]
So my rule is actually a little different. It's called `no-next-next` and it tries to prevent people from calling `next()` twice in middleware on accident. Trying to follow this, it sounds like it's not terribly related, but here's what I'm currently using for that (like I said, it's only half working):
https://gist.github.com/xjamundx/1d7ccfe4589f4cb306c4

**Comment by puzrin** [comment]
@xjamundx sorry, i'm busy with other projects and have no plans to implement this rule in nearest future. I share all ideas, which can be useful for community or need further discussion, but implement only those which are critical for me. This one is interesting but not critical.

**Comment by xjamundx** [comment]
This bit us again at work. Will try to take a stab at it :)

**Comment by xjamundx** [comment]
I've gotten to the place where these are all correctly handled (per my understanding)

```js
- √ "function a(err) { if (err) return callback(err); }",
- √ "function a(err) { if (err) return callback(err); callback(); }",
- √ "function a(err) { if (err) { return callback(err); } callback(); }",
- X "function a(err) { if (err) { callback (err); } }"
```
Let's talk through several more examples to make sure I understand the expected behavior:

```js
// correct and good behavior
- √ "function x(err) { if (err) { callback(); return; } }"
- √ "function x(err) { if (err) { callback(); return; } return callback(); }"
- √ "function x(err) { if (err) { return callback(); } else { return callback(); } }"
- √ "function x(err) { if (err) return callback(); else return callback(); }"

// known bad behavior which we ignore
- √ "function x(err) { if (err) { setTimeout(callback 0); } callback(); }" // callback() called twice

// good behavior that we warn against to simplify the rule
- X "function x(err) { if (err) { callback() } else { callback() } }" // actually correct, just confusing
```

**Comment by nzakas** [comment]
Looks good!

**Comment by gcochard** [comment]
:+1: I strongly support this rule!

What about the following:

I foresee this when we want to force the callback to be async so we don't [release zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony). We are pairing the callback with return, but the first is inside of `process.nextTick` so it *will* be called twice.
```js
function x(err) {
  if(err){
    process.nextTick(function(){
      return callback(err);
    });
  }
  return callback();
}
```

**Comment by nzakas** [comment]
Are you asking if the rule would warn about this case? I'm not really sure from your description.

**Comment by gcochard** [comment]
Yes, I would like the rule to warn about this case. Should have been more clear about that.

**Comment by nzakas** [comment]
What is it warning about, exactly? Each callback has a return.

**Comment by gcochard** [comment]
It violates (IMHO) the spirit of the rule, which is to keep a callback from being called twice. Putting the callback call inside an anonymous function causes the return to be in the wrong place. It should paired with `process.nextTick` so that each code path which calls a callback is a terminal path.

**Comment by nzakas** [comment]
I think it's a stretch to include that case in this rule. I'd vote for keeping to the original description.

**Comment by gcochard** [comment]
I'm okay with that. Just make a note that deferred callbacks may not be flagged by this rule.

**Comment by xjamundx** [comment]
`"callback-return": [0, ["callback", "cb", "next"]]` seem like a good name and reasonable defaults?

**Comment by xjamundx** [comment]
I thought about turning this on by default in the node environment and ran into this...

```
Validating Makefile.js

./Makefile.js
  721:12  error  Expected return with your async function  callback-return

✖ 1 problem (1 error, 0 warnings)
```

```
lib/config-initializer.js
  32:8  error  Expected return with your async function  callback-return

lib/util/traverse.js
  61:16  error  Expected return with your async function  callback-return
  81:8   error  Expected return with your async function  callback-return
```

**Comment by xjamundx** [comment]
I guess I'll just remove that, because we're getting rules automatically being turned on anyway....

**Comment by lo1tuma** [comment]
> `"callback-return": [0, ["callback", "cb", "next"]]` seem like a good name and reasonable defaults?

What about `done`, which is often used in asynchronous tests (e.g. with mocha).
